File: ios/Stockbook/Features/Sell/SellScreen.swift
import SwiftUI

/// The bill being written: a form, with a product picker behind one button.
///
/// **A form, not a cart.** A bill here is a number, a date, somebody and a
/// figure — the paper book was written first, so the total is already known and
/// rebuilding it product by product to arrive at it is work for nothing. Saying
/// what was sold is optional, and the only thing it buys is the shelf moving.
///
/// Add items is therefore a genuine mode, entered and left by tapping, rather
/// than the derived state it used to be. An empty cart no longer means the
/// picker: it is what the ordinary bill in this shop looks like.
struct SellScreen: View {
    @Environment(StockbookStore.self) private var store
    @Environment(AppRouter.self) private var router
    @Environment(Cart.self) private var cart

    private var products: [Product] { store.products }

    @State private var query = ""
    /// Set by "Add items", cleared by "Done adding". On the router rather than in
    /// here because the shell draws the tab bar and must not stack it under the
    /// picker's own bottom bar.
    private var browsing: Bool { router.pickingProducts }

    private var showsPicker: Bool {
        browsing || !query.isBlank
    }

    private var matches: [Product] { store.products(matching: query) }

    var body: some View {
        VStack(spacing: 0) {
            ScreenHeader(title: Loc.newBill, bottomPadding: 10) {
                Text(cartCountLabel)
                    .font(NocturneType.inter(12))
                    .foregroundStyle(Nocturne.neutral500)
            }

            if showsPicker, !pickerHint.isEmpty {
                Text(pickerHint)
                    .nocturneText(.meta)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.horizontal, Metrics.screenPadding)
                    .padding(.bottom, 8)
            }

            // Which of the two is showing is derived, so the swap is the only
            // signal that a tap changed anything. It fades rather than cuts.
            Group {
                if showsPicker {
                    // The search box belongs to the picker rather than to the
                    // screen: a form for typing a figure should not open with a
                    // box asking for a product name.
                    NocturneField(placeholder: Loc.addAProductPlaceholder, text: $query, fontSize: 14.5)
                        .padding(.horizontal, Metrics.screenPadding)
                        .padding(.bottom, 10)

                    ProductPicker(
                        products: matches,
                        hasAnyProducts: !products.isEmpty,
                        query: query.trimmed,
                        onPick: add(_:),
                        onAddProduct: { router.openNewProduct() },
                        onDoneAdding: closePicker
                    )
                } else {
                    CartView(
                        onBrowse: openPicker,
                        onSave: save
                    )
                }
            }
            .transition(.opacity)
            .motion(Motion.screen, value: showsPicker)
        }
        // Leaving Sell puts the picker away. It used to happen for free, when
        // this was a `@State` that died with the screen.
        .onDisappear { router.pickingProducts = false }
    }

    private var cartCountLabel: String {
        cart.isEmpty ? "" : Loc.lines(cart.lines.count)
    }

    private var pickerHint: String {
        guard !products.isEmpty else { return "" }
        if !query.isBlank { return Loc.matchingQuery(query.trimmed) }
        return Loc.allProductsHint(products.count)
    }

    // MARK: Actions

    /// Adds a piece and clears the search, so the next product can be typed
    /// straight away without reaching for the clear button.
    private func add(_ product: Product) {
        cart.add(product)
        query = ""
    }

    private func openPicker() {
        router.pickingProducts = true
        query = ""
    }

    private func closePicker() {
        router.pickingProducts = false
        query = ""
    }

    /// Saves a new bill, or rewrites the one being corrected.
    ///
    /// `amount` is passed either way. The store ignores it when there are lines —
    /// one rule, in one place, rather than a screen deciding which of two figures
    /// is the real one.
    ///
    /// A correction closes onto the corrected document rather than the receipt
    /// overlay: the owner came from that sheet and is checking the change landed,
    /// not confirming a sale that has just happened.
    private func save() {
        if let number = cart.editing {
            guard let corrected = store.updateBill(
                number: number,
                lines: cart.draftLines,
                customer: cart.customer,
                paid: cart.paidForStorage,
                amount: cart.typedAmount,
                createdAt: cart.soldAt,
                invoiceNo: cart.invoiceNo
            ) else { return }

            cart.clear()
            closePicker()
            router.tab = .book
            router.openBill(corrected)
            return
        }

        guard let bill = store.saveBill(
            lines: cart.draftLines,
            customer: cart.customer,
            paid: cart.paidForStorage,
            amount: cart.typedAmount,
            createdAt: cart.soldAt,
            invoiceNo: cart.invoiceNo
        ) else { return }

        cart.clear()
        closePicker()
        router.receipt = bill
    }
}

/// The browse-and-search list. Tapping a row adds one piece at the product's
/// current selling price; tapping one already in the cart increments it.
private struct ProductPicker: View {
    let products: [Product]
    let hasAnyProducts: Bool
    let query: String
    let onPick: (Product) -> Void
    let onAddProduct: () -> Void
    let onDoneAdding: () -> Void

    @Environment(Cart.self) private var cart
    @Environment(\.currency) private var currency
    @Environment(\.bottomSafeInset) private var bottomInset

    var body: some View {
        VStack(spacing: 0) {
            ScrollView {
                LazyVStack(spacing: 5) {
                    if products.isEmpty {
                        EmptyStateBox(
                            message: emptyMessage,
                            actionTitle: Loc.addAProduct,
                            action: onAddProduct
                        )
                        .padding(.top, 8)
                    }

                    ForEach(products) { product in
                        Button {
                            onPick(product)
                        } label: {
                            HStack(spacing: 10) {
                                Text(product.name)
                                    .font(NocturneType.inter(14))
                                    .frame(maxWidth: .infinity, alignment: .leading)
                                    .lineLimit(1)

                                // Browsing keeps you on this list, so without a
                                // mark here a tap produces no visible result and
                                // the same item gets added twice.
                                let onBill = cart.quantity(forProduct: product.uid)
                                if onBill > 0 {
                                    HStack(spacing: 3) {
                                        Glyph(Icon.confirm, size: 11)
                                        Text("\(onBill)")
                                            .font(NocturneType.inter(11.5))
                                            .contentTransition(.numericText())
                                    }
                                    .foregroundStyle(Nocturne.accent)
                                    // The list does not move when a row is
                                    // tapped, so this mark is the whole of the
                                    // feedback. It arrives with a pop.
                                    .transition(.scale(scale: 0.4).combined(with: .opacity))
                                }

                                Text(Loc.stockLabel(product.stock))
                                    .nocturneText(.meta)
                                    .lineLimit(1)
                                Text(Money.text(product.price, in: currency))
                                    .font(NocturneType.inter(14))
                                    .foregroundStyle(Nocturne.accent400)
                            }
                            .padding(.horizontal, 12)
                            .padding(.vertical, 11)
                            .frame(maxWidth: .infinity)
                            .background(Nocturne.surface, in: RoundedRectangle(cornerRadius: Metrics.controlRadius, style: .continuous))
                            .contentShape(Rectangle())
                        }
                        .buttonStyle(.plain)
                        .motion(Motion.pop, value: cart.quantity(forProduct: product.uid))
                        .accessibilityLabel(
                            cart.quantity(forProduct: product.uid) > 0
                                ? Loc.onBillAccessibility(
                                    name: product.name,
                                    quantity: cart.quantity(forProduct: product.uid)
                                  )
                                : product.name
                        )
                    }
                }
                .padding(.horizontal, Metrics.screenPadding)
                .padding(.bottom, 18)
            }
            .scrollDismissesKeyboard(.interactively)

            // With a cart in progress the tab bar is hidden, so this footer is
            // the only way back to it.
            if !cart.isEmpty {
                HStack(spacing: 12) {
                    VStack(alignment: .leading, spacing: 0) {
                        Text(Loc.lines(cart.lines.count))
                            .nocturneText(.meta)
                        Text(Money.text(cart.total, in: currency))
                            .font(NocturneType.inter(19, .medium))
                            .rollingNumber(cart.total)
                    }
                    Spacer(minLength: 12)
                    Button(Loc.doneAdding, action: onDoneAdding)
                        .buttonStyle(PrimaryButtonStyle(height: 44))
                }
                .padding(.horizontal, Metrics.screenPadding)
                .padding(.top, 10)
                .padding(.bottom, max(bottomInset, 24))
                .background(Nocturne.surface)
                .overlay(alignment: .top) {
                    Rectangle().fill(Nocturne.neutral800).frame(height: 1)
                }
            }
        }
    }

    private var emptyMessage: String {
        hasAnyProducts ? Loc.noProductMatches(query) : Loc.noProductsYet
    }
}

File: android/app/src/main/kotlin/com/stockbook/app/feature/sell/SellScreen.kt
package com.stockbook.app.feature.sell

import androidx.compose.animation.Crossfade
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.stockbook.app.AppRouter
import com.stockbook.app.design.EmptyStateBox
import com.stockbook.app.design.GhostButton
import com.stockbook.app.design.Glyph
import com.stockbook.app.design.Icon
import com.stockbook.app.design.Metrics
import com.stockbook.app.design.Motion
import com.stockbook.app.design.Nocturne
import com.stockbook.app.design.NocturneField
import com.stockbook.app.design.NocturneType
import com.stockbook.app.design.PrimaryButton
import com.stockbook.app.design.ScreenHeader
import com.stockbook.app.design.card
import com.stockbook.core.model.Product
import com.stockbook.core.model.ShopState
import com.stockbook.core.money.Money
import com.stockbook.core.store.StockbookStore
import com.stockbook.core.text.Strings

/**
 * The billing flow: the bill form, with the product picker one tap behind it.
 *
 * The form is home now, not the picker. A bill in this shop is a number, a date,
 * somebody and a figure — the picker exists for the minority of bills that also
 * say what was sold, so it is reached from "Add items" and left again by "Done
 * adding". The form never disappears from under an owner who has typed into it:
 * everything they entered is still there when they come back.
 *
 * It is also where a bill gets **corrected**. `router.editingBill` is what says
 * so, and the form is the same one either way — a second screen for editing is a
 * second screen to keep in step with this one, and it would lose that race.
 */
@Composable
fun SellScreen(
    state: ShopState,
    store: StockbookStore,
    router: AppRouter,
    cart: Cart,
    strings: Strings,
    modifier: Modifier = Modifier
) {
    var query by remember { mutableStateOf("") }
    // Set by "Add items", cleared by "Done adding". On the router rather than in
    // here because the shell draws the tab bar and must not stack it under the
    // picker's own bottom bar.
    val browsing = router.pickingProducts
    // Null for a new bill. The cart was filled from this one before the tab
    // changed, so everything below draws it without knowing where it came from.
    val editing = router.editingBill

    // Leaving Sell puts the picker away. It used to happen for free, when this
    // was a `remember` that died with the screen.
    DisposableEffect(Unit) { onDispose { router.pickingProducts = false } }

    val currency = state.settings.currency
    // An empty cart no longer means the picker: a bill with nothing on it is the
    // ordinary bill here, and dropping the owner into a product list to write one
    // would be the app asking a question the paper already answered.
    val showsPicker = browsing || query.isNotBlank()
    val matches = remember(state.products, query) { store.productsMatching(query) }

    Column(modifier = modifier.fillMaxWidth()) {
        ScreenHeader(
            // Which bill is being corrected, said in the heading: the form below
            // is identical to the one for a new bill, so this line is the whole
            // of what tells the owner they are changing 1024 rather than writing
            // 1025.
            kicker = if (editing != null) strings.editBill else null,
            title = editing?.reference(strings) ?: strings.newBill,
            bottomPadding = 10.dp
        ) {
            if (editing != null) {
                // The way out of a correction without making one. Without it an
                // owner who tapped Edit by mistake has no way back to a blank
                // form except by saving the bill they did not mean to touch.
                GhostButton(
                    strings.cancel,
                    onClick = {
                        cart.clear()
                        router.closeBillEditing()
                        query = ""
                    },
                    tint = Nocturne.neutral500
                )
            } else if (!cart.isEmpty) {
                // Only where there is a count to give. "Empty" beside a form with
                // a customer and a figure already in it describes the item list
                // and reads as a description of the bill.
                Text(
                    strings.lines(cart.lines.size),
                    style = NocturneType.inter(12.0),
                    color = Nocturne.neutral500
                )
            }
        }

        Crossfade(targetState = showsPicker, animationSpec = Motion.screenSpec, label = "sell") { picking ->
            if (picking) {
                Column(modifier = Modifier.fillMaxWidth()) {
                    if (state.products.isNotEmpty()) {
                        Text(
                            if (query.isNotBlank()) strings.matchingQuery(query.trim())
                            else strings.allProductsHint(state.products.size),
                            style = NocturneType.meta,
                            color = Nocturne.neutral500,
                            modifier = Modifier
                                .padding(horizontal = Metrics.screenPadding)
                                .padding(bottom = 8.dp)
                        )
                    }

                    // The search box belongs to the picker rather than to the
                    // screen: a form for typing a total should not open with a
                    // box asking for a product name.
                    NocturneField(
                        value = query,
                        onValueChange = { query = it },
                        placeholder = strings.addAProductPlaceholder,
                        fontSize = 14.5,
                        modifier = Modifier
                            .padding(horizontal = Metrics.screenPadding)
                            .padding(bottom = 10.dp)
                    )

                    ProductPicker(
                        products = matches,
                        hasAnyProducts = state.products.isNotEmpty(),
                        query = query.trim(),
                        cart = cart,
                        currency = currency,
                        strings = strings,
                        onPick = { product ->
                            cart.add(product)
                            query = ""
                        },
                        onAddProduct = { router.openNewProduct() },
                        onDoneAdding = {
                            router.pickingProducts = false
                            query = ""
                        },
                        modifier = Modifier.weight(1f)
                    )
                }
            } else {
                CartView(
                    cart = cart,
                    state = state,
                    store = store,
                    currency = currency,
                    strings = strings,
                    onBrowse = {
                        router.pickingProducts = true
                        query = ""
                    },
                    editing = editing,
                    onSave = {
                        // The same figures either way, into whichever of the two
                        // this is. `updateBill` moves the shelf by the difference
                        // between the old bill and the new one, which is why an
                        // edit is one call rather than a removal and a save.
                        val bill = if (editing != null) {
                            store.updateBill(
                                number = editing.number,
                                lines = cart.draftLines,
                                customer = cart.customer,
                                paid = cart.paidForStorage,
                                amount = cart.typedAmount,
                                createdAt = cart.soldAt,
                                invoiceNo = cart.invoiceNo
                            )
                        } else {
                            store.saveBill(
                                lines = cart.draftLines,
                                customer = cart.customer,
                                paid = cart.paidForStorage,
                                // Passed whether or not there are lines. The store
                                // ignores it when there are — one rule, in one
                                // place, rather than a screen deciding which
                                // figure is real.
                                amount = cart.typedAmount,
                                createdAt = cart.soldAt,
                                invoiceNo = cart.invoiceNo
                            )
                        }
                        if (bill != null) {
                            cart.clear()
                            query = ""
                            if (editing != null) {
                                // Back to the list the bill was opened from, where
                                // the corrected row is the confirmation. A receipt
                                // here would announce a sale that did not happen.
                                router.closeBillEditing()
                            } else {
                                router.pickingProducts = false
                                router.receipt = bill
                            }
                        }
                    }
                )
            }
        }
    }
}

/**
 * The browse-and-search list. Tapping a row adds one piece at the product's
 * current selling price; tapping one already on the bill increments it.
 */
@Composable
private fun ProductPicker(
    products: List<Product>,
    hasAnyProducts: Boolean,
    query: String,
    cart: Cart,
    currency: com.stockbook.core.model.Currency,
    strings: Strings,
    onPick: (Product) -> Unit,
    onAddProduct: () -> Unit,
    onDoneAdding: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(modifier = modifier.fillMaxWidth()) {
        LazyColumn(
            modifier = Modifier.weight(1f),
            contentPadding = PaddingValues(
                start = Metrics.screenPadding,
                end = Metrics.screenPadding,
                bottom = 18.dp
            )
        ) {
            if (products.isEmpty()) {
                item {
                    EmptyStateBox(
                        message = if (hasAnyProducts) strings.noProductMatches(query) else strings.noProductsYet,
                        actionTitle = strings.addAProduct,
                        onAction = onAddProduct,
                        modifier = Modifier.padding(top = 8.dp)
                    )
                }
            }

            items(products, key = { it.uid }) { product ->
                val onBill = cart.quantity(product.uid)
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(bottom = 5.dp)
                        .card(Metrics.controlRadius)
                        .clickable { onPick(product) }
                        .padding(horizontal = 12.dp, vertical = 11.dp)
                ) {
                    Text(
                        product.name,
                        style = NocturneType.inter(14.0),
                        color = Nocturne.text,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.weight(1f)
                    )
                    // The list does not move when a row is tapped, so this mark
                    // is the whole of the feedback.
                    if (onBill > 0) {
                        Glyph(Icon.confirm, size = 11.dp, tint = Nocturne.accent)
                        Spacer(Modifier.width(3.dp))
                        Text("$onBill", style = NocturneType.inter(11.5), color = Nocturne.accent)
                        Spacer(Modifier.width(10.dp))
                    }
                    Text(
                        strings.stockLabel(product.stock),
                        style = NocturneType.meta,
                        color = Nocturne.neutral500
                    )
                    Spacer(Modifier.width(10.dp))
                    Text(
                        Money.text(product.price, currency),
                        style = NocturneType.inter(14.0),
                        color = Nocturne.accent400
                    )
                }
            }
        }

        // Always, not only with something on the bill: the picker is a place the
        // owner stepped into from the form, and an owner who steps in and changes
        // their mind must have the same way out as one who added six things.
        Column(modifier = Modifier.fillMaxWidth().background(Nocturne.surface)) {
            Box(Modifier.fillMaxWidth().height(Metrics.hairline).background(Nocturne.neutral800))
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = Metrics.screenPadding, vertical = 10.dp)
                    .navigationBarsPadding()
            ) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        if (cart.isEmpty) strings.cartEmpty else strings.lines(cart.lines.size),
                        style = NocturneType.meta,
                        color = Nocturne.neutral500
                    )
                    // The line sum, and only where there are lines to sum. With
                    // none picked yet the figure here would be whatever was typed
                    // into the amount box, sitting under the word "empty".
                    if (!cart.isEmpty) {
                        Text(
                            Money.text(cart.total, currency),
                            style = NocturneType.bigNumber(19.0),
                            color = Nocturne.text
                        )
                    }
                }
                PrimaryButton(strings.doneAdding, onClick = onDoneAdding, height = 44.dp)
            }
        }
    }
}
